Duplicati/Library/Backend/Telegram/Extensions/TelegramClientExtensions.cs

```diff
@@ -18,31 +45,87 @@ public static bool IsReallyConnected(this TelegramClient client)
+            var sender = GetMtProtoSender(client);
 
             if (sender == null)
             {
                 return false;
             }
 
-            var transportFieldInfo = typeof(TelegramClient).GetField("transport", _bindingFlags);
-            var transportField = (TcpTransport)transportFieldInfo.GetValue(client);
+            var tcpTransport = client.GetTcpTransport();
 
-            if (transportField == null || transportField.IsConnected == false)
+            if (tcpTransport == null || tcpTransport.IsConnected == false)
             {
                 return false;
             }
 
-            var tcpClientFieldInfo = typeof(TcpTransport).GetField("tcpClient", _bindingFlags);
-            var tcpClient = (TcpClient)tcpClientFieldInfo.GetValue(transportField);
+            var tcpClient = tcpTransport.GetTcpClient();
 
             if (tcpClient == null || tcpClient.Connected == false)
             {
                 return false;
             }
-            
+
             return true;
         }
+
+        private static TcpClientConnectionHandler GetTcpClientConnectionHandler(this TelegramClient client)
+        {
+            if (client == null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+
+            var handlerFieldInfo = typeof(TelegramClient).GetField("handler", _bindingFlags);
+            var handler = (TcpClientConnectionHandler)handlerFieldInfo.GetValue(client);
+
+            return handler;
+        }
+
+        private static TcpClient GetTcpClient(this TcpTransport tcpTransport)
+        {
+            if (tcpTransport == null)
+            {
+                throw new ArgumentNullException(nameof(tcpTransport));
+            }
+
+            var tcpClientFieldInfo = typeof(TcpTransport).GetField("tcpClient", _bindingFlags);
+            var tcpClient = (TcpClient)tcpClientFieldInfo.GetValue(tcpTransport);
+
+            return tcpClient;
+        }
+
+        private static FieldInfo GetTcpTransportFieldInfo(this TelegramClient client)
+        {
+            var transportFieldInfo = typeof(TelegramClient).GetField("transport", _bindingFlags);
+
+            return transportFieldInfo;
+        }
+
+        private static TcpTransport GetTcpTransport(this TelegramClient client)
+        {
+            if (client == null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+
+            var transportFieldInfo = client.GetTcpTransportFieldInfo();
+            var oldTransportField = (TcpTransport)transportFieldInfo.GetValue(client);
+
+            return oldTransportField;
+        }
+
+        private static MtProtoSender GetMtProtoSender(this TelegramClient client)
+        {
+            if (client == null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+
+            var senderFieldInfo = typeof(TelegramClient).GetField("sender", _bindingFlags);
```

Comment:
Is the use of reflection necessary?  I'm worried it makes us sensitive to internal refactorings of `TelegramClient`.

Reply:
There was only one way to achieve the solution yet.
Later the reflection will be removed and the underlying library will be changed to [TgSharp](https://github.com/nblockchain/TgSharp)

Now they do not support 2-factor authentication so we can't use it yet.